### PR TITLE
Statusbar

### PR DESCRIPTION
**Pull request by OceanWolf**
Now we have renamed `ToolbarBase` to `ToolContainerBase`, the message area no longer seems to fit here.  Even more so when we consider that we **can** have more then one toolbar, say one on the left and one on the right, etc.

Of course this doesn't force backends to have a separate status bar, as their `Toolbar` class can simply inherit from `StatusbarBase` as well as `ToolContainerBase`, though of course it makes sense for the backends to separate these two features out as we do here with Gtk3.

### CODE REVIEW

**Comment by fariza** [comment]
I understand your point, but at the same time we want to keep it simple. 
You could easily create a dedicated `ToolContainer` that only implements the message area and the other containers that don't implement it.

**Comment by OceanWolf** [comment]
Yes, but if I wanted to have two `ToolbarGTK3`s in the future, then both would contain the message area.  This seems overkill to me.

In terms of usage, it makes no difference as the events abstract that away.  In terms of code simplicity, while it does add a new GTK element, it makes the GTK3 code shorter by several lines by not having to fiddle around with the complexities of the adding the statusbar into the toolbar.  (Note I left some un-needed code in there for doing tool-positions, which means less code to add for tool-positions).

So with tool positions the layout currently looks like this:
- Window VBox
  - FigureCanvas
  - VBox
    - HBox
      - Toolbar Group
      - Separator
      - Toolbar Group
    - HBox
      - Label

But now this simplifies to:
- Window VBox
  - FigureCanvas
  - HBox
    - Toolbar Group
    - Separator
    - Toolbar Group
  - Statusbar

**Comment by fariza** [comment]
Does this brings, the same kind of simplification for other backends?

**Comment by OceanWolf** [comment]
I have no idea, I don't know the various backends that well, though I would imagine so, statusbars represent a pretty core feature to applications.  As I say though, one doesn't have to make any changes to the backend if they don't want to/can't, i.e. if gtk3 didn't support statusbar, then I could have just done the following, leaving the rest of the code intact:

``` python
class ToolbarGTK3(ToolContainerBase, StatusbarBase, Gtk.Box):
    def __init__(...):
        StatusbarBase.__init__(self, navigation)
```

Also by doing it like this, it gives StatusbarBase more room to grow as a separate entity post MEP-22 -- if programmers feel they want to add more base features to it -- without the clutter of the `ToolContainer` code getting in the way.

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -3734,3 +3719,24 @@ def remove_toolitem(self, name):
         """
 
         raise NotImplementedError
+
+class StatusbarBase(object):
```
Comment (fariza) [inline lib/matplotlib/backend_bases.py:3724]: Pep8: 2 blank lines

File: lib/matplotlib/backends/backend_gtk3.py [hunk 2]
```diff
@@ -413,16 +414,22 @@ def __init__(self, canvas, num):
...
             self.toolbar.add_tools(toolbar_tools)
+            self.statusbar = StatusbarGTK3(self.navigation)
 
-        # calculate size for window
-        w = int (self.canvas.figure.bbox.width)
-        h = int (self.canvas.figure.bbox.height)
+            self.statusbar.show()
+            self.vbox.pack_end(self.statusbar, False, False, 0)
+            size_request = self.statusbar.size_request()
+            h += size_request.height
```
Comment (fariza) [inline lib/matplotlib/backends/backend_gtk3.py]: Add a separator between the toolbar and statusbar, just to make it look more conventional.

**Comment by fariza** [comment]
Right now, the text is on the right, do we keep it that way or we change it?

**Comment by OceanWolf** [comment]
I don't know, what do you think?  Before didn't the text only go on the right because the tools were on the same line and on the left?

**Comment by fariza** [comment]
Ok, leave the text on the left, we'll see later on. Just add the separator and we are ready to go (I think)

**Comment by OceanWolf** [comment]
Okay, patches made, moved into sub-function, as having the same bit of code written thrice a bit much me thinks.

**Comment by OceanWolf** [comment]
bump

**Comment by fariza** [comment]
Hopefully tonight. Kids and wife beats python

**Comment by OceanWolf** [comment]
Ooh, yes, completely agree! :).  In the mean time... I shall see if I find time to do an alternate gcf refactor off master akin to my suggestions in that thread, hopefully that will make this navigation stuff easier to implement **cleanly**  _cough_!  Unless your mult-fig-manager will deal with that.  I kind of have the idea that your multi-fig-manager will become the only manager, with an option whether to put into tabs, or new window.

Should I give it a go?

**Comment by OceanWolf** [comment]
P,S. I meant to say that I completely understand.  I just thought as you were commenting on my other PRs that perhaps you had missed that I had fixed those gripes.

As a heads up, I have started on an alternate refactor of Gcf out of navigation, and it looks nice so far (if I do say so myself), about halfway/two-thirds, I should get a fully refactored version done in the next few hours, so that the fine-tuning for specific backends can begun (only concentrating on `gtk3` at the moment).  This has ramifications for MEP-22 and also this specific PR, and I mean ramifications in a nice way, a lot less backend-specific coding to get MEP-22 working on all backends :).

**Comment by OceanWolf** [comment]
Just started the backend refactor of Qt4, and I notice that they have the toolbar on the top with a statusbar at the bottom, with text aligned left, so going down this route definitely doesn't matter :wink:

On an aside, they have an extra tool "Edit curves lines and axis parameters".  Will make for an interesting conversion when it comes to migrating MEP22 to Qt4  :yum:

**Comment by OceanWolf** [comment]
Working now on Wx and that uses a statusbar, a custom mpl status, so you just need to make that class extend this and perfect, no extra work :).

**Pull request merged**